extensions/ql-vscode/src/helpers.ts

```diff
@@ -464,3 +457,72 @@ export class CachedOperation<U> {
     }
   }
 }
+
+
+
+/**
+ * The following functions al heuristically determine metadata about databases.
+ */
+
+const dbSchemeToLanguage = {
+  'semmlecode.javascript.dbscheme': 'javascript',
+  'semmlecode.cpp.dbscheme': 'cpp',
+  'semmlecode.dbscheme': 'java',
+  'semmlecode.python.dbscheme': 'pyhton',
+  'semmlecode.csharp.dbscheme': 'csharp',
+  'go.dbscheme': 'go',
+  'ruby.dbscheme': 'ruby'
+};
+
+/**
+ * Returns the initial contents for an empty query, based on the language of the selected
+ * databse.
+ *
+ * First try to get the contents text based on language. if that fails, try to get based on
+ * dbscheme. Otherwise return no import statement.
+ *
+ * @param language the database language or empty string if unknown
+ * @param dbscheme path to the dbscheme file
+ *
+ * @returns an import and empty select statement appropriate for the selected language
+ */
+export function getInitialQueryContents(language: string, dbscheme: string) {
+  if (!language) {
+    const dbschemeBase = path.basename(dbscheme) as keyof typeof dbSchemeToLanguage;
+    language = dbSchemeToLanguage[dbschemeBase];
+  }
+
+  return language
+    ? `import ${language}\n\nselect ""`
+    : 'select ""';
+}
+
+export async function isLikelyDatabaseRoot(maybeRoot: string) {
+  const [a, b, c] = (await Promise.all([
+    // databases can have either .dbinfo or codeql-database.yml.
+    fs.pathExists(path.join(maybeRoot, '.dbinfo')),
+    fs.pathExists(path.join(maybeRoot, 'codeql-database.yml')),
+
+    // they *must* have a db-{language} folder
+    glob('db-*/', { cwd: maybeRoot })
+  ]));
+
+  return !!((a || b) && c);
+}
+
+export function isLikelyDbLanguageFolder(dbPath: string) {
+  return !!path.basename(dbPath).startsWith('db-');
+}
+
+export async function getPrimaryLanguage(root: string) {
```

Comment:
Instead of parsing the config file, make a call to `codeql resolve database --format json` and get the first value in the `languages` field.

Reply:
Agreed. Especially because the open-sourced VSCode extension is our showcase for the Right way to do things.

Reply:
That will also mean that the extension will not need its own `dbSchemeToLanguage` lookup table, since the `codeql resolve database` now already tries a dbscheme-based heuristic for language detection.

Reply:
The reason I chose not to do that is that currently, there is no cli available on `openDatabase`. Also, making a filesystem check will be faster than starting up a jvm to get this information.

I could certainly stuff a reference to the cli into the constructor. But I'd prefer not to do that for the reasons above unless the approach I'm taking would produce incorrect results.

Reply:
I would lean towards considering "people who write their own IDE integrations for CodeQL think this is the right thing to do, and proceed to bake half-understood internal knowledge of database internals into their own work" to be an _incorrect result_. :-)

On the other hand, I don't have the insight to appreciate the possible problems in dealing with "there is no cli available on `openDatabase`". Why is that?

Reply:
The `cli` is a singleton that gets passed around when needed. The `DatabaseManager` does not hold a reference to the `cli`. I would need to change that in order to call a cli command here. It's not a big change, but it is an architectural one.

Also, I still need `dbSchemeToLanguage` as a fallback in case the database was not created by a modern CLI. I will add a comment around using that so it's known that this is not recommended.

So, the question is, should I be getting the database label from the `codeql-database.yml` or from `database resolve`? I'm not sure what your comment was on this or about the `dbSchemeToLanguage`.